base_location/models/res_city_zip.py

```diff
@@ -0,0 +1,47 @@
+# Copyright 2016 Nicolas Bessi, Camptocamp SA
+# Copyright 2018 Aitor Bouzas <aitor.bouzas@adaptivecity.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class BetterZip(models.Model):
+    """City/locations completion object"""
+
+    _name = "res.city.zip"
+    _description = __doc__
+    _order = "name asc"
+
+    name = fields.Char('ZIP')
+    city_id = fields.Many2one(
+        'res.city',
+        'City',
+        required=True,
+    )
+    latitude = fields.Float()
+    longitude = fields.Float()
+
+    @api.multi
+    @api.depends('name', 'city', 'state_id', 'country_id')
+    def name_get(self):
+        result = []
+        for rec in self:
+            name = []
+            if rec.name:
+                name.append('%(name)s' % {'name': rec.name})
+            name.append('%(name)s' % {'name': rec.city_id.name})
+            if rec.city_id.state_id:
+                name.append('%(name)s' % {'name': rec.city_id.state_id.name})
+            if rec.city_id.country_id:
+                name.append('%(name)s' % {'name': rec.city_id.country_id.name})
+            result.append((rec.id, ", ".join(name)))
```

Comment:
I'm thinking that maybe you can add a computed stored field depending on these values for not overloading each `name_get` call, having to access 3 tables. This will rise the entries addition (for example, through `base_location_geonames_import`), but will speed up location search, which is the most habitual task. Being called `display_name`, it will be automatically done without overwriting this method.